src/main/kotlin/me/moeszyslak/polly/services/MacroService.kt

```diff
@@ -25,36 +27,96 @@ import me.moeszyslak.polly.data.MacroStore
         val channelId = channel?.id?.value ?: ""
         val name = name_raw.toLowerCase()
-        val category = category_raw.toLowerCase()
 
-        if (name in discord.commands.map { it.names }.flatten().map { it.toLowerCase() }) {
+        val parent = store.forGuildWithMacro(guild, name, channelId) { _, parent, _ ->
+            parent
+        } ?: run {
+            event.respond("Cannot find a macro by that name. If it is a channel specific macro you need to provide the channel as well.")
+            return
+        }
+
+        event.respond {
+            title = "Macro - ${name}"
+            color = discord.configuration.theme
+            description = "```${parent.contents}```"
+            field {
+                this.name = "Macro Name"
+                value = "```fix\n${parent.name}```"
+                inline = true
+            }
+            field {
+                this.name = "Aliases"
+                value = "```fix\n${parent.aliases.joinToString("\n")}```"
+                inline = true
+            }
+            field {
+                this.name = "Uses"
+                value = "`${parent.uses}`"
+            }
+            field {
+                this.name = "Category"
+                value = "`${parent.category}`"
+                inline = true
+            }
+            field {
+                this.name = "Channel"
+                value = if (parent.channel ?: "" == "") "`Global Macro`"
+                else event.guild.getChannel(Snowflake(parent.channel!!)).mention
+                inline = true
+            }
+        }
+    }
+
+    fun addMacro(guild: GuildId, nameRaw: String, categoryRaw: String, channel: TextChannel?, contents: String): String {
+        val channelId = channel?.id?.value ?: ""
+        val (name, aliases) = nameRaw
+                .toLowerCase()
+                .split("|")
+                .distinct()
+                .let { it.first() to it.drop(1) }
+        val fullName = "$name#$channelId"
+        val category = categoryRaw.toLowerCase()
+
+        val cmds = discord.commands.map { it.names }.flatten().map { it.toLowerCase() }
+        if (name in cmds || aliases.any { it in cmds }) {
             return "A command with that name already exists."
         }
 
-        val result = store.forGuild(guild) {
-            it.putIfAbsent("$name#$channelId", Macro(name, contents, channelId, category))
+        val result = store.forGuild(guild) { macros ->
+            if (macros.containsKey(fullName)) return@forGuild null
+            if (aliases.any { macros.containsKey("$it#$channelId") }) return@forGuild null
+
+            val parent = Macro(name, null, aliases, contents, channelId, category)
+            macros[fullName] = parent
+
+            aliases.forEach {
+                macros["$it#$channelId"] = parent.copy(name = it, parent = fullName, aliases = listOf())
```

Comment:
The parent-child thing looks unnecessarily complicated and it's also duplicating aliases in the macro store and persisted data. Can we do something simpler? For example, we keep only "real" macros in the main macro store and then have a different non-persistent map of aliases (including the canonical name) to the canonical. This does the constant time lookup for an alias as well as avoid some bit of hassle managing the macro copies in the store. After any update to the map, you just regenerate this, thus you only need one small piece of code and everything else almost stays the same. This alias map can either go in the MacroStore (but marked Transient so it's not persistent) or in the MacroService.